### ISSUE TITLE
Any way to make a heatmap?

### ISSUE THREAD

**Issue by guysoft** [philipperemy/keract#12]
Hey,
So is there a way to write something similar to this: https://jacobgil.github.io/deeplearning/class-activation-maps

I'd appreciate a pointer how to modify ``display_activations`` to be able to write a heat map and would gladly PR. 
I am not sure how it merges all the feature layers together.

**Comment by philipperemy** [comment]
Hey @guysoft ,

Thanks for your comment! I spent ~1 hour to implement it. Here's what I got:

Please correct it and improve it!

```python
from __future__ import print_function

from glob import glob

import keras.backend as K
import numpy as np
from scipy.misc import imresize

from data import get_mnist_data

# What this script does:
# - define the model
# - if no checkpoints are detected:
#   - train the model
#   - save the best model in checkpoints/
# - load the model from the best checkpoint
# - read the activations

if __name__ == '__main__':

    checkpoints = glob('examples/checkpoints/*.h5')
    # pip3 install natsort
    from natsort import natsorted

    from keras.models import load_model

    if len(checkpoints) > 0:

        checkpoints = natsorted(checkpoints)
        assert len(checkpoints) != 0, 'No checkpoints found.'
        checkpoint_file = checkpoints[-1]
        print('Loading [{}]'.format(checkpoint_file))
        model = load_model(checkpoint_file)

        model.compile(optimizer='adam',
                      loss='categorical_crossentropy',
                      metrics=['accuracy'])
        print(model.summary())

        image_id = 33
        conv_output = model.layers[1].output
        x_train, y_train, x_test, y_test = get_mnist_data()
        conv_output_grads = K.gradients(model.total_loss, conv_output)[0]
        inputs = model._feed_inputs + model._feed_targets + model._feed_sample_weights + [K.learning_phase()]
        gradient_func = K.function(inputs, [conv_output_grads, conv_output])
        output_grad, output_activations = gradient_func([[x_train[image_id]], [y_train[image_id]], [1], False])
        mul_grad_act = np.sum(np.abs(output_grad * output_activations), axis=-1).squeeze()

        import matplotlib.pyplot as plt

        plt.imshow(x_train[image_id].squeeze())
        plt.show()

        plt.imshow(imresize(mul_grad_act, (28, 28)))
        plt.show()

    else:
        pass
```

I guess it's possible

**Comment by guysoft** [comment]
Hey, just updating I didn't forget this, just got a bunch of stuff that entered my pipeline and this has been pushed backed, just FYI

**Comment by philipperemy** [comment]
@guysoft ok np I'll leave it open then!

**Comment by guysoft** [comment]
Ok, managed to get back to this.

I can get the example to work with MINST. Note you need to:
1. clone keract,
2. place the file above in the ``examples`` folder
3. change ``checkpoints = glob('examples/checkpoints/*.h5')`` to ``checkpoints = glob('examples/checkpoints/*.h5')``.
4. run ``mnist.py`` to get checkpoints
5. run file above.

However I can't seem to adapt it to my network. Need to find what to assign in ``inputs =`` , `` K.gradients`` (what is set in the example to ``model.total_loss``).

I did manage to use this:
https://github.com/keras-team/keras/blob/master/examples/class_activation_maps.py

**Comment by qwertpi** [comment]
I've got heatmaps working (albeit only with yellow for low activation and dark red for high activations) if you want to take a look at my pull request (#26)

**Comment by philipperemy** [comment]
Awesome!

**Comment by qwertpi** [comment]
> I've got heatmaps working (albeit only with yellow for low activation and dark red for high activations) if you want to take a look at my pull request (#26)

I think I've got blue-dark red colours working, an example color bar can be found here https://drive.google.com/open?id=1KV_NuKnarpjVWTkzBNhAfVHBXjQAefq3

**Comment by philipperemy** [comment]
Good job! Are you fairly confident in the implementation?

**Comment by qwertpi** [comment]
I'd like to say yes but it's probably best if you check it over

**Comment by philipperemy** [comment]
@qwertpi I did a first check and it seems to be correct!
I'll do a second check tomorrow then will merge!

**Comment by philipperemy** [comment]
@qwertpi merged!
https://github.com/philipperemy/keract/commit/72f022704dc7d2f0387e0a4ef1ee6eac5f2e552a
Do you need a new version of the library or can it wait for other changes?
